### ISSUE TITLE
Proposal: `github-release` type

### ISSUE THREAD

**Issue by steiza** [package-url/purl-spec#299]
This issue is to suggest a `github-release` purl type, discuss the motivation, and list some possible alternatives.

### Motivation

At GitHub we're working on [some improvements to GitHub Releases](https://github.com/github/roadmap/issues/943) including the ability to make a release immutable (after which you couldn't add, change, or remove any of the assets on the release).

The immutability would be externally auditable by [publishing an in-toto release attestation](https://github.com/in-toto/attestation/blob/main/spec/predicates/release.md), which includes a purl. As an aside, the in-toto release attestation is a formalization of the [npm publish attestation](https://github.com/npm/attestation/tree/main/specs/publish/v0.1), which has been out for over a year and is working well for over 10,000 different npm packages.

npm publish attestations use the npm purl type, but GitHub Releases doesn't have a purl type.

### Proposal

Add a purl type `github-release`. For example, https://github.com/advanced-security/gh-sbom/releases/tag/v0.0.10 would look like:

```
pkg:github-release/advanced-security/gh-sbom@v0.0.10
```

To be explicit, here's a short definition for each of those purl fields:

- **type**: `github-release`
- **namespace**: GitHub "owner" (e.g. a user or an organization)
- **name**: repository name
- **version**: the version string of the GitHub Release

There is some weirdness here to call out:
- In the purl spec the version is optional, but a `github-release` without a version doesn't make much sense.
- Today GitHub Releases are tied to a git tag, but we're thinking of ways to decouple GitHub Releases from git tags, because git tags are mutable.

### Alternative 1: use existing `github` purl type

The problem with this is that the purl version is ambiguous. Today the purl spec says the version can refer to a commit hash or a git tag. We'd like it to refer to the release version string, but there's no requirement that the git tag and the release version string match (also the git tag is mutable). For example:

```
pkg:github/advanced-security/gh-sbom@v0.0.10
```

Could refer to https://github.com/advanced-security/gh-sbom/tree/v0.0.10, which is mutable, or (if we don't have a separate `github-release` type) https://github.com/advanced-security/gh-sbom/releases/tag/v0.0.10, which could be made immutable.

### Alternative 2: use `generic` purl type

The problem with this approach is that it's very easy to have purl name collisions. Maybe this would look something like:

```
pkg:generic/gh-sbom@v0.0.10?download_url=https://github.com/advanced-security/gh-sbom/releases/tag/v0.0.10
```

Any other package named `gh-sbom` (a fork, or an unrelated project) would end up with a similar-looking purl.

### Conclusion

Feedback welcome! If folks think this sounds good, I can write up a more formal pull request.

**Comment by jhutchings1** [comment]
I'm a fan of the proposal to use a new type here as you propose, @steiza.

**Comment by crt-31** [comment]
> pkg:generic/gh-sbom@v0.0.10?download_url=https://github.com/advanced-security/gh-sbom/releases/tag/v0.0.10
Any other package named gh-sbom (a fork, or an unrelated project) would end up with a similar-looking purl.

Just trying to understand: why is this an issue if they are similar. Wouldn't they be different purls since the download_url is different?

**Comment by jkowalleck** [comment]
like the idea.

lets think about how a specific file from a release could be pointed out.

lets use this as an example: https://github.com/package-url/packageurl-python/releases/tag/v0.16.0
- how would i express that i need the asset <https://github.com/package-url/packageurl-python/releases/download/v0.16.0/packageurl_python-0.16.0.tar.gz>? 
  answer is probably: well-known qualifier "file_name"
- how would i express that i need a certain file/asset?
  answer is probably: subpath

**Comment by pombredanne** [comment]
> @jaimergp it turns out that the feature I'm working (the one which inspired the need for this new purl type) is to make releases immutable. When used in that context, the repo and tag should be sufficient to uniquely identify the release. Perhaps an optional `checksum` qualifier could be added to account for the current state of mutable releases?

_Originally posted by @bdehamer in https://github.com/package-url/purl-spec/issues/418#issuecomment-2777064551_
            
hum ... my knee jerk reaction is mixed feelings about this type! Let me think about this more and come with pros and cons.

@bdehamer if this is immutable, why separate this from the github type? Would there be a way to keep it there? After all, a release is also a tag?

**Comment by matt-phylum** [comment]
Releases have tags but they are more than tags and can contain arbitrary additional files.

**Comment by bdehamer** [comment]
@pombredanne yeah, GitHub Releases today are tightly coupled to git tags today and we certainly could use the existing `github` type to identify a release. That said, we see a couple benefits to having a dedicated `github-release` type:

* Being able document clear constraints/semantics for PURL components that are specific to GitHub releases (e.g. specifying that the `file_name` qualifier can be used to identify a specific release asset).
* Provides an unambiguous way to identify a release even if GitHub Releases become decoupled from git tags in the future.
